src/stats.ml

```diff
@@ -45,3 +45,117 @@ let dump () =
 let enable () =
   enabled := true;
   at_exit dump
+
+module Catapult = struct
+  type active_state = Pervasives.out_channel * Format.formatter
+
+  type state =
+    | Disabled
+    | Path of string
+    | Active of active_state
+
+  let state = ref Disabled
+
+  let dispose () =
+    match !state with
+    | Disabled -> ()
+    | Path _ -> ()
+    | Active (channel, _) ->
+      Pervasives.output_string channel "]\n";
+      Pervasives.flush channel;
+      Pervasives.close_out channel
+
+  let enable path =
+    state := Path path;
+    at_exit dispose
+
+  let fmt () =
+    match !state with
+    | Disabled -> None
+    | Active (_, fmt) ->
+      Format.pp_print_string fmt ",";
+      Some fmt
+    | Path p ->
+      let channel = Pervasives.open_out_gen [Open_append; Open_creat] 0o666 p in
+      let fmt = Format.formatter_of_out_channel channel in
+      Format.pp_print_string fmt "[";
+      state := Active (channel, fmt);
+      Some fmt
+
+  let printf format_string =
+    match fmt () with
+    | Some fmt -> Format.fprintf fmt (format_string ^^ "\n%!")
+    | None -> Format.ifprintf Format.std_formatter format_string
+
+  let color_of_name = function
+    | "ocamlc.opt" -> "thread_state_uninterruptible"
+    | "ocamlopt.opt" -> "thread_state_running"
+    | "ocamldep.opt" -> "thread_state_runnable"
+    | "ocamlmklib.opt" -> "thread_state_unknown"
+    | "ocamllex.opt" -> "thread_state_sleeping"
```

Comment:
It's a bit unfortunate to hardcode the `.opt` extensions here. As we might be using the bytecode versions of these on some systems. Perhaps we should include the bytecode versions as well, or have the `Context.t` to use the exact binary names.

Reply:
Sure, I'll do that. I should add that there's only a finite set of colors, so the "terrible", "bad" etc names are not meaningful here.